app/src/split/bluetooth/service.c

```diff
@@ -24,12 +26,59 @@ LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
 static uint8_t num_of_positions = ZMK_KEYMAP_LEN;
 static uint8_t position_state[POS_STATE_LEN];
 
+#define BEHAVIOR_LABEL_LEN 12
+
+static struct zmk_split_run_behavior_payload behavior_run_payload;
+
 static ssize_t split_svc_pos_state(struct bt_conn *conn, const struct bt_gatt_attr *attrs,
                                    void *buf, uint16_t len, uint16_t offset) {
     return bt_gatt_attr_read(conn, attrs, buf, len, offset, &position_state,
                              sizeof(position_state));
 }
 
+static ssize_t split_svc_run_behavior(struct bt_conn *conn, const struct bt_gatt_attr *attrs,
+                                      const void *buf, uint16_t len, uint16_t offset,
+                                      uint8_t flags) {
+    struct zmk_split_run_behavior_payload *payload = attrs->user_data;
+    uint16_t end_addr = offset + len;
+
+    LOG_DBG("offset %d len %d", offset, len);
+
+    if (end_addr > sizeof(struct zmk_split_run_behavior_payload)) {
+        return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+    }
+
+    memcpy(payload + offset, buf, len);
+
+    // We run if:
+    // 1: We've gotten all the position/state/param data.
+    // 2: We have a null terminated string for the behavior device label.
+    if ((end_addr > sizeof(struct zmk_split_run_behavior_data)) &&
+        payload->behavior_dev[end_addr - sizeof(struct zmk_split_run_behavior_data)] == '\0') {
```

Comment:
I think using `offsetof` would make it more obvious what this is doing, and it would handle changes to the layout of `zmk_split_run_behavior_payload` without needing to update this.

Might look something like this:
```c
const size_t behavior_dev_offset = offsetof(struct zmk_split_run_behavior_payload, behavior_dev);
if (end_addr > behavior_dev_offset && payload->behavior_dev[end_addr - behavior_dev_offset] == '\0') {
    ...
```

Reply:
Indeed. Refactored.